File: Yandex/Practicum/Algorightms_and_Data_Structures/2_sprint/H.Скобочная_последовательность/solution.py
import sys


def solution(braces: str) -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    stack: list[str] = []
    valid_open_braces: dict[str, str] = {
        ')': '(', ']': '[', '}': '{',
    }
    is_valid: bool = True
    for brace in braces:
        if brace in '([{':
            stack.append(brace)
        else:
            if not stack or stack[-1] != valid_open_braces[brace]:
                is_valid = False
                break
            stack.pop()
    sys.stdout.write(["False", "True"][is_valid and not stack] + '\n')

if __name__ == '__main__':
    braces: str = sys.stdin.readline().rstrip()
    solution(braces)

File: Yandex/Practicum/Algorightms_and_Data_Structures/2_sprint/H.Скобочная_последовательность/solution.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <stack>

void solution(std::string& braces) {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    std::stack<char> stack;
    bool isValid = true;
    std::unordered_map<char, char> validOpenBraces = {
        {')', '('}, {']', '['}, {'}', '{'}
    };
    for (char brace : braces) {
        if (brace == '(' || brace == '[' || brace == '{') {
            stack.push(brace);
        } else {
            if (!stack.size() || stack.top() != validOpenBraces[brace]) {
                isValid = false;
                break;
            }
            stack.pop();
        }
    }
    std::cout << (isValid && !stack.size()? "True" : "False") << "\n";
}

int main() {
    std::string braces;
    std::cin >> braces;
    solution(braces);
}